[adr] stop service while waiting for cert (#181)
canonical/observability: decision-records/2024-06-04--service-to-stop-while-waiting-for-cert.md

**Date:** 2024-06-04<br/>
**Author:** @sed-i

## Context and Problem Statement
When an admin forms a tls relation, there is a time window in which:

> 1. the juju model tells us we must operate with TLS enabled; however

Reviewer: ```suggestion
1. the juju model tells us we must operate with TLS enabled (the desired/planned state); however
```

> 2. we haven't received the server cert yet.

Reviewer: ```suggestion
2. we haven't received the server cert yet (actual, transient state).
```

as I mention below, this isn't probably _just about_ the cert missing, but about whether all preconditions are met for enabling tls, including:
- the cert is on disk (charm and workload(s))
- update-ca-certificates --fresh has run and exited 0
- charm has privkey

As a result we had some code ordering issues (e.g. #414):

> - Should we update the config file with the tls config, which includes path to

Reviewer: if yes, so what? if no, so what? I'd like to see the tradeoffs more explicitly worded in this point and the next

Author: Updated.

  the cert, before we got the cert, only to perfectly reflect the juju model?
- Should we wait with the tls config and allow the workload to continue to
  operate in plain http, in contradiction to the juju model?

> - How should we determine within the charm if "TLS is enabled"?

Reviewer: I think the issue is one of semantics more than anything: what does it mean for the charm to be tls-enabled? 
at what point in time does, precisely, the charm have all the pieces of data it needs (in the right place) for it to start talking tls with the other end? And in those intermediate states when some of the data isn't quite there yet, but the charm is already aware of the 'desired state' and knows that we're going towards it, what should the charm do?

> This problem is common to many charms.

Reviewer: probably, all charms that support tls. 
Should we mention that this is a gap in the juju model and we're here only because of that, so hopefully at some point this adr will be obsoleted by juju itself?

Author: You mean that TLS isn't ubiquitous in juju? Seems far fetched to talk about it.

Reviewer: I mean that tls is an ubiquitous problem in juju, which juju should model but doesn't yet

Reviewer: but yea, too general perhaps?

## Decision

> A server shall NOT keep using plain http while waiting for a cert. The service

Reviewer: I'd phrase this more generally, it's not just about waiting for the cert, it's about the cert being on disk (workload AND charm): what if pebble is temporarily offline or there's a glitch in the matrix and secret-get returns rubbish?
In general, I'd say, "the charm shall NOT keep using plain http during any of the transient intermediate states"

Author: Updated.

must be stopped until the cert is in place.

> The charm shall have a "Waiting" status during that period (subject to other

Reviewer: ```suggestion
The charm shall surface this state to the admin by setting a "Waiting" status during that period (subject to other
```

Author: Either phrasing is probably correct, depending on [what you mean by "charm"](https://discourse.charmhub.io/t/what-do-we-mean-when-we-say-charm/8603).

status priorities).

> We also distinguish between several substates:

Reviewer: this is good!

- TLS disabled: a tls-certificate is not in place, no certs on disk, no tls
  section in app config, app is running in plain HTTP mode.
- TLS enabled: a tls-certificate relation is present.
- Cert ready: we have a cert in relation data.
- TLS config ready: the cert that is in relation data was written to disk, and
  app config has a corresponding tls section.

> - TLS active: the app is running with the new config.

Reviewer: I'd separate this more clearly from the 'inconsistent transient states' above, as this is the final happy state we're aiming to reach

    [*] --> TLSDisabled : install

    TLSDisabled: TLS disabled
    note right of TLSDisabled
        No TLS relation
        No certs on disk
        No TLS section in app config
        App running in plain HTTP mode
    end note
    TLSDisabled --> TLSEnabled : tls-relation-joined

    OUT: Service outage
    state OUT {
        TLSEnabled: TLS enabled
        note left of TLSEnabled
            TLS relation present
            CSR sent, waiting for cert, or cert expired or revoked
        end note
        TLSEnabled --> CertReady: tls-relation-changed
        CertReady: Cert ready
        note left of CertReady
            A live cert is in relation data
        end note
        CertReady --> TLSReady
        TLSReady: TLS config ready
        note right of TLSReady
            The cert that is in relation data was written to disk
            App config has a corresponding tls section
        end note
        TLSReady --> TLSActive
    }

    TLSActive: TLS active
    note right of TLSActive
        App restarted and operates in TLS
    end note
    TLSActive --> TLSActive : cert update

    TLSActive --> TLSEnabled : cert revoked, cert expired
```

## Benefits

> - No need to worry if it takes the cert a long time to arrive.

Reviewer: or if the charm remains in any other inconsistent transient state for 'long'

Author: The other inconsistent states are not a new concern so I think it's out of scope.
I.e. if pebble push or update cert fails is not specific to this ADR.

> - We stick to the juju model (no model departure): if admin says HTTPS, then no

Reviewer: what does (no model departure) mean here?

  more plain HTTP.

## Disadvantages
- We may incur data loss if the cert takes so long to arrive that clients begin
  to drop data due to back-pressure.

## Alternatives considered

> - Keep operating in plain http until we receive the cert.

Reviewer: do we need to mention the key reasons this was rejected?

Reviewer: fair enough
